lib/Controller/FolderApiV2Controller.php

```diff
@@ -0,0 +1,129 @@
+<?php
+/**
+ * Nextcloud - News
+ *
+ * This file is licensed under the Affero General Public License version 3 or
+ * later. See the COPYING file.
+ *
+ * @author    Paul Tirk <paultirk@paultirk.com>
+ * @copyright 2020 Paul Tirk
+ */
+
+namespace OCA\News\Controller;
+
+use \OCP\IRequest;
+use \OCP\IUserSession;
+use \OCP\AppFramework\Http;
+
+use \OCA\News\Service\FolderServiceV2;
+use \OCA\News\Service\ItemServiceV2;
+use \OCA\News\Service\Exceptions\ServiceNotFoundException;
+use \OCA\News\Service\Exceptions\ServiceConflictException;
+use \OCA\News\Service\Exceptions\ServiceValidationException;
+
+class FolderApiV2Controller extends ApiController
+{
+    use ApiV2ResponseTrait;
+
+    private $folderService;
+    private $itemService;
+
+    public function __construct(
+        IRequest $request,
+        IUserSession $userSession,
+        FolderServiceV2 $folderService,
+        ItemServiceV2 $itemService
+    ) {
+        parent::__construct($request, $userSession);
+
+        $this->folderService = $folderService;
+        $this->itemService = $itemService;
+    }
+
+    /**
+     * @NoAdminRequired
+     * @NoCSRFRequired
+     * @CORS
+     *
+     * @param string $name
+     * @return array|mixed|\OCP\AppFramework\Http\JSONResponse
+     */
+    public function createFolder($name)
+    {
+        try {
+            $this->folderService->purgeDeleted($this->getUserId(), false);
+            $responseData = $this->serialize(
+                $this->folderService->create($this->getUserId(), $name)
+            );
+            return $this->response([
+                'folder' => $responseData
+            ]);
+        } catch (ServiceValidationException $ex) {
+            return $this->errorResponse($ex, Http::STATUS_BAD_REQUEST);
+        } catch (ServiceConflictException $ex) {
+            $responseData = $this->serialize(
+                $this->folderService->findByName($this->getUserId(), $name)
+            );
+            return $this->response([
+                'folder' => $responseData
+            ], Http::STATUS_CONFLICT);
+        }
+    }
+
+    /**
+     * @NoAdminRequired
+     * @NoCSRFRequired
+     * @CORS
+     * @param int    $folderId
+     * @param string $name
+     * @return array|\OCP\AppFramework\Http\JSONResponse
+     */
+    public function updateFolder($folderId, $name)
+    {
+        $response = null;
+        try {
+            $response = $this->folderService->rename($this->getUserId(), $folderId, $name);
+        } catch (ServiceValidationException $ex) {
+            return $this->errorResponse($ex, Http::STATUS_UNPROCESSABLE_ENTITY);
+        } catch (ServiceConflictException $ex) {
+            $responseData = $this->serialize(
+                $this->folderService->findByName($this->getUserId(), $name)
+            );
+            return $this->response(
+                [
+                    'folder' => $responseData
+                ],
```

Comment:
I'm not sure if this ever happens, but if it does you're practically leaking user data. I'd prefer it didn't have the response data

Reply:
The draft of the API says to return the existing folder if it is tried to create. Personally I don't see why there could not be two folders with the same name.. from a database point of view it wouldn't matter..

Reply:
There's also nothing stopping that in the service layer. I think this might be the draft document showing its age.